features/FEATURE_BLE/source/generic/GenericGap.cpp

```diff
@@ -850,7 +850,7 @@ void GenericGap::on_read_phy(
 )
 {
     ble_error_t status = BLE_ERROR_NONE;
-    if (pal::hci_error_code_t::SUCCESS != hci_status) {
+    if (pal::hci_error_code_t(pal::hci_error_code_t::SUCCESS) != hci_status) {
```

Comment:
It seems like putting in an explicit conversion potentially defeats the point of the type-safety, although I'm not sure as of this C++98 version. Conceptually, at least in newer C++, you'd be activating an explicit conversion, which is a bit like forcing a cast.

I see that just swapping the test to the sane way round clears the problem too:

    if (hci_status != pal::hci_error_code_t::SUCCESS) {

Haven't quite understood why yet, but I'll seize on any justification to get people to write their comparisons in the sane order, so I like that.

Reply:
Thanks for the report @kjbracey-arm ; that's interesting. As documented in SafeEnum, the main goal was to avoid scope collision between enum of different types so you don't convert a peer_address_type_t into an own_address_type_t as they have slightly different semantics. 

Unfortunately we opened a hole in the type safety with comparison operators operating on the base class and not the derived one. Fortunately it can be fixed by rewriting comparison operators to use `Target` (the derived class) instead of `SafeEnum` ([live version](https://godbolt.org/z/fi5fpO)). Do you think that's enough ? If so I can issue a patch today.

Note that MSVC is also impacted with operand orders. I need to understand why.

Note: As you expressed it, once C++14 is here we can just use `enum class`

Reply:
Actually `if (4 != hci_status)` works on every compiler. clang only complains if the first parameter is an enum type. Issue raised with clang: https://bugs.llvm.org/show_bug.cgi?id=40329

For now, I think just swap the order - that will retain type safety when/if `hci_error_code_t` becomes `enum class : uint8_t`.

Reply:
@pan-, ah, the proper-safety is simpler than I thought - I was overthinking it.   That looks good to me. Go for that.

(But we could still put the comparisons the right way round, on pure taste grounds)

Reply:
Actually, your rewrite doesn't actually solve the order problem - still doesn't manage the reverse comparison in clang. So that's just a separate type-safety fix.

Reply:
yes, it is just a separate fix for type safety. I'm trying to understand why clang and msvc refuses to accept the expressions `hci_error_code_t::SUCCESS == error`. Otherwise I agree with you; the order of operand in this code must be swapped. We don't need Yoda notation.

Reply:
> clang folks agreed it was a bug, and it has now been fixed upstream - you can now see it working for "clang (trunk)" in compiler explorer.

@kjbracey-arm 💯  !

Reply:
Fix visible here for any compiler geeks: https://github.com/llvm-mirror/clang/commit/a7d7e428cf6497d79dc3a557586cf9c7fe7b4839   (this example is reduced to a test case).

The code modified was clearly marked "FIXME" :)